feat: allow changing slides via a CarouselRef; add more className props

hsource opened the pull request:
### Description

1. We want to advance/change the current slide of the carousel from a parent outside the carousel.
2. To customize styling of the slider list and individual slides, we don't want to rely on the slider-list and similar classNames, as they're internal implementation details. Instead, we let users pass in their own classNames
3. Make Carousel height update when only the slide itself changes without any props changing by using ResizeObserver (when available)

#### Type of Change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

#### Refs

Added new Storybook story that also includes contains buttons to change slide via the new refs:

![Ref controls](https://github.com/FormidableLabs/nuka-carousel/assets/2937410/d2d9e7d2-cde6-438e-9f7d-38e5a117c0a2)

#### Resizing on children change

![Change heights](https://github.com/FormidableLabs/nuka-carousel/assets/2937410/8c381ae0-c02f-406b-ad90-b30ad210427e)

### Checklist: (Feel free to delete this section upon completion)

- [X] My code follows the style guidelines of this project (I have run `yarn lint`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes (I have run `yarn test` and `yarn test-e2e`)
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

hsource commented on src-v5/carousel.tsx:
```diff
@@ -15,476 +26,486 @@ import {
-    isMounted.current = true;
-    return () => {
-      isMounted.current = false;
+const Carousel = React.forwardRef<CarouselRef, CarouselProps>(
```
I changed very few lines. Make sure to review this with "Hide whitespace" checked.

![image](https://user-images.githubusercontent.com/2937410/168988603-a6019ce3-faae-4722-b434-b74fb9526a21.png)

hsource commented:
I'll rebase this for merge after the others are pulled, since this changes a lot of indentation

hsource commented:
@carlos-kelly @gksander - this PR lets parent components control the carousel with a ref. To make merging easier without rebasing, I just added 2 commits beyond what #909 has. Take a look after #909 if this seems useful to you! We'd love to get back onto mainline and not have to run with patches

gksander commented:
@hsource we just merged two of your other PRs. Can you rebase with/merge in `main` and push up to this branch so we can see a better diff, and we'll get this reviewed soon! 🙏 Thank you for all of the awesome contributions!

hsource commented:
@gksander - thanks! I've rebased and did a quick read of it just now. I think it's ready to review!

huyphan1999 commented:
Hi, there is any progress on this PR? 
I'm meeting the issue with adaptive height. The carousel is stuck in zero height because adaptive height is calculated before the image load

hsource commented:
@gksander Just a heads up - I rebased this, self-reviewed the code, and verified that the Storybook stories still work well. I updated the gifs of me testing the feature in the description.

@carloskelly13 too in case the ownership of this has changed! I saw you've been reviewing/pulling PR's lately.

Thanks again for the great package!

hsource commented on packages/nuka/src/carousel.tsx:
```diff
@@ -175,11 +186,11 @@ export const Carousel = React.forwardRef<HTMLDivElement, CarouselProps>(
 
     const forwardedRef = useForwardRef<HTMLDivElement>(ref);
-    const carouselRef = forwardedRef || defaultCarouselRef;
+    const carouselDivRef = forwardedRef || defaultCarouselDivRef;
```
I renamed the ref pointing to the carousel's div to `carouselDivRef` to avoid confusing with the new `carouselRef` prop, which is now the external carousel ref.

I kept the forwarded ref as the ref to the div to maintain backwards compatibility.

hsource commented on packages/nuka/src/slide.tsx:
```diff
@@ -119,47 +121,71 @@ const Slide = ({
-      className={`slide${currentSlideClass}${
-        typeOfSlide ? ` ${typeOfSlide}` : ''
-      }${isFullyVisible ? ' slide-visible' : ''}`}
+      className={[
+        'slide',
+        isCurrentSlide && isVisible && 'slide-current',
+        typeOfSlide,
+        isFullyVisible && 'slide-visible',
+        slideClassName,
+      ]
+        .filter(Boolean)
+        .join(' ')}
```
I didn't change any of the existing classes. I changed the className to this format since it was a bit more readable and easier to maintain
